### PR TITLE
Added api, impl, central, local, release repos for minimal workspace.

### PR DESCRIPTION
**Pull request by scottslewis**
Also added maven central bundle org.osgi.annotation.bundle:2.0.0 so that java service wizards function properly.

### CODE REVIEW

**Comment by scottslewis** [comment]
@pkriens  please approve workflow for build of this min workspace template to proceed

**Comment by pkriens** [comment]
- I created the min workspace to have a clean starting point. I really would like to have a single repository that is absolutely minimal and is not based on the MavenBndRepository for complexity. 
- The Pom version is  convenient for us because it is common between maven & bndtools but I really dislike it. The `bnd mbr` command has lots of useful functions to update the Maven Bnd Repositoroes automatically. 
- Having the absolute minimal is important for a lot of people to start from the most minimum.
- We should also adopt OSGi version 8 and not seven. I also think the local repositories should not be there.
- We could make the default a repository larger

**Comment by scottslewis** [comment]
> * I created the min workspace to have a clean starting point. I really would like to have a single repository that is absolutely minimal and is not based on the MavenBndRepository for complexity.
> 
>     * The Pom version is  convenient for us because it is common between maven & bndtools but I really dislike it. The `bnd mbr` command has lots of useful functions to update the Maven Bnd Repositoroes automatically.
> 
>     * Having the absolute minimal is important for a lot of people to start from the most minimum.
> 
>     * We should also adopt OSGi version 8 and not seven. I also think the local repositories should not be there.
> 
>     * We could make the default a repository larger

Hi @pkriens .

I suggest you tweak things to your liking and submit a new pr.   The *only* thing that the java service templates actually require is the org.osgi.annotation.bundle:2.0.0 in central.maven...which of course requires the maven central (or some) repo with org.osgi.annotation.bundle:2.0.0...which of course requires the maven central repo entry in build.bnd file.

So please feel free to delete anything/everything except the repo to your liking in build.bnd.  I'm of the personal view that the min workspace is currently so minimal as to be useless....and so more is definitely needed, but I'll leave to you and others to decide what else needs to be build.bnd to make it useful in additions to supporting the use of the java service template (i.e. the annotation.bundle).

In short, I don't want to submit multiple prs to guess what's in your notion of minimal, but I do want to have the minimal workspace template support the new java service template...that seems pretty minimal to me.

**Comment by pkriens** [comment]
I agree that I should do some work here but it would be nice to have an absolute minimal workspace so feedback will be appreciated.

**Comment by scottslewis** [comment]
Ok here are my thoughts on minimal:

> * I created the min workspace to have a clean starting point. I really would like to have a single repository that is absolutely minimal and is not based on the MavenBndRepository for complexity.

Why a single repo?    I don't see having a single repo as particularly valuable for the initial dev user.  What I would consider to be more complex is to require the new user to have to go about finding and 'installing' the repos/bundles needed to do anything at all (like use the service template).

I agree that local and release repos could probably be left out.   I think Templates repo is valuable, though as a placeholder...and maybe for examples/tutorials, etc.

> 
> * The Pom version is  convenient for us because it is common between maven & bndtools but I really dislike it. The `bnd mbr` command has lots of useful functions to update the Maven Bnd Repositoroes automatically.

I think in today's world that it's sort of expected that maven central is considered a minimal resource.   Not great, IMHO, but here we are.

If the org.osgi.annotation.bundle:2.0.0 can be made available in min workspace from somewhere other than maven central then I'm fine with doing so, but I do think that maven central should at least be there...so people can see that it's easily accessible in bndtools.   That will make many of them more comfortable IMHO.

> 
> * Having the absolute minimal is important for a lot of people to start from the most minimum.

I don't really understand this point.   Yes I get the meaning of minimal, but I think that wrt tooling minimal usually means 'minimal workspace that allows me to do something useful' (implied:  without having to configure the tools themselves).  I don't think it typically means minimal in the java 'hello world' sense....especially for something like modules/osgi.

> 
> * We should also adopt OSGi version 8 and not seven. I also think the local repositories should not be there.

I have no problem with osgi 8.   Is everything settled wrt spec/api/impl/framework/availability in repos, etc?   If so then +1.

RE the local repo.   This is judgment call IMHO...on the other side for 'local':   the presence of the Local repo does provide devs with some clearer idea of how bundles can/could be added to repos without using MC or someone else's repo...e.g. drag and drop jars onto local repo.   This is an important thing to divulge to devs...and having 'Local' appear in the repositories view is kind of a nice breadcrumb IMHO and much preferable to doc explanation.

> 
> * We could make the default a repository larger

As per above, given the repositories view, I don't think it adds a tremendous amount of complexity to have more than one repo in repo view...even for a minimal workspace.   OTOH I think the notion of having just one repo might be nice for the 'minimalist' dev but it also might confuse some users to assume that there is only one repo *possible*.

**Comment by pkriens** [comment]
> Ok here are my thoughts on minimal:
> 
> > * I created the min workspace to have a clean starting point. I really would like to have a single repository that is absolutely minimal and is not based on the MavenBndRepository for complexity.
> 
> Why a single repo? I don't see having a single repo as particularly valuable for the initial dev user. What I would consider to be more complex is to require the new user to have to go about finding and 'installing' the repos/bundles needed to do anything at all (like use the service template).

Simplicity & the aggregate paradox. When people start something new they want something that works out of the but box. However, once it works they want imho to be able to inspect the parts and understand these. When you get a large blob of things you do not understand this is a lot harder.

The second thing is that in a aggregating world like OSGi it is a lot easier to start minimal and add parts then to start big. Inevitably there will be incompatibilities with new parts and trying to resolve that in a large setup takes a lot of effort.

I am a sicker for <=> less is more. A lot of complexity is caused by quantity.

Last but not least. A minimum repo is easier to maintain. Large repos get hairy quickly.

> 
> I agree that local and release repos could probably be left out. I think Templates repo is valuable, though as a placeholder...and maybe for examples/tutorials, etc.
> 
> > * The Pom version is  convenient for us because it is common between maven & bndtools but I really dislike it. The `bnd mbr` command has lots of useful functions to update the Maven Bnd Repositoroes automatically.
> 
> I think in today's world that it's sort of expected that maven central is considered a minimal resource. Not great, IMHO, but here we are.

The Maven Bnd Repository also uses Maven Central but its format is a lot more simple to use. It is just a line with the G:A:V. Entries can be updated from the Repositories view to the latest version and the `bnd mbr` command has ways to verify it can update all to the latest version. The G:A:V can be set in a separate file or even in the bnd file itself and macros are supported (although this can interfere with the auto update facility).

> 
> If the org.osgi.annotation.bundle:2.0.0 can be made available in min workspace from somewhere other than maven central then I'm fine with doing so, but I do think that maven central should at least be there...so people can see that it's easily accessible in bndtools. That will make many of them more comfortable IMHO.
> 
> > * Having the absolute minimal is important for a lot of people to start from the most minimum.
> 
> I don't really understand this point. Yes I get the meaning of minimal, but I think that wrt tooling minimal usually means 'minimal workspace that allows me to do something useful' (implied: without having to configure the tools themselves). I don't think it typically means minimal in the java 'hello world' sense....especially for something like modules/osgi.

It is subjective but I hate large

> 
> > * We should also adopt OSGi version 8 and not seven. I also think the local repositories should not be there.
> 
> I have no problem with osgi 8. Is everything settled wrt spec/api/impl/framework/availability in repos, etc? If so then +1.
> 
> RE the local repo. This is judgment call IMHO...on the other side for 'local': the presence of the Local repo does provide devs with some clearer idea of how bundles can/could be added to repos without using MC or someone else's repo...e.g. drag and drop jars onto local repo. This is an important thing to divulge to devs...and having 'Local' appear in the repositories view is kind of a nice breadcrumb IMHO and much preferable to doc explanation.
> 
> > * We could make the default a repository larger
> 
> As per above, given the repositories view, I don't think it adds a tremendous amount of complexity to have more than one repo in repo view...even for a minimal workspace. OTOH I think the notion of having just one repo might be nice for the 'minimalist' dev but it also might confuse some users to assume that there is only one repo _possible_.

As said, it is subjective but I assigned this to me and will make a proposal.

**Comment by scottslewis** [comment]
<stuff deleted>
> 
> Simplicity & the aggregate paradox. When people start something new they want something that works out of the but box. However, once it works they want imho to be able to inspect the parts and understand these. When you get a large blob of things you do not understand this is a lot harder.

My view is that there is no paradox.   Yes people want something useful out of box.  But they don't necessarily want to understand the parts...rather in my view they want help with 'progressive disclosure'...i.e. to *be able* to understand the parts of tooling/languages that they are interested in/need when they need them.   That's why having progressive disclosure is useful...things can be 'hidden' (e.g. multiple repos) behind UI (repositories view) and only exposed if/when needed (change add/remove what repos are available).   Note thinking about it this way opens up possibility for additional/simplifying UI...e.g. a 'add repo' button+dialgo to repositories view.

> 
> The second thing is that in a aggregating world like OSGi it is a lot easier to start minimal and add parts then to start big. Inevitably there will be incompatibilities with new parts and trying to resolve that in a large setup takes a lot of effort.

I agree with you in general, but it can be taken too far...e.g. an empty workspace template.   I don't understand why it has to be '1 repo' or 'large setup'.   Why not 'just right'?

> 
> I am a sicker for <=> less is more. A lot of complexity is caused by quantity.

I get that you are a sicker for less is more, but it's not your usage of bndtools I'm concerned about ;-).

> 
> Last but not least. A minimum repo is easier to maintain. Large repos get hairy quickly.

So again:  what about 2 or three repos with a small number of initial (i.e. small jar references)?   We are talking goldilocks here :).

<stuff deleted>
> 
> The Maven Bnd Repository also uses Maven Central but its format is a lot more simple to use. It is just a line with the G:A:V. Entries can be updated from the Repositories view to the latest version and the `bnd mbr` command has ways to verify it can update all to the latest version. The G:A:V can be set in a separate file or even in the bnd file itself and macros are supported (although this can interfere with the auto update facility).

I'm just saying you should meet people/devs where they are.   I would recommend staying away from yourself as a user (or in fact any of the existing bndtools developers).   

For example, have you ever watched a brand new bndtools user try to do something with bndtools for the first time (e.g. define a simple/hello world osgi service)...without (anyone) providing any input/direction to that new user?  

<stuff deleted>
> > 
> > 
> > I don't really understand this point. Yes I get the meaning of minimal, but I think that wrt tooling minimal usually means 'minimal workspace that allows me to do something useful' (implied: without having to configure the tools themselves). I don't think it typically means minimal in the java 'hello world' sense....especially for something like modules/osgi.
> 
> It is subjective but I hate large

Large is subjective, however.   What seems to you to be large and unwieldy (e.g. maven central as a bundle resource) may be completely familiar (and so not perceived as large and complex if coming from java ee world).   OTOH, something that seems very familiar and simple to you/me (e.g. osgi services) may not be perceived as such by those same developers.

<stuff deleted>

> > As per above, given the repositories view, I don't think it adds a tremendous amount of complexity to have more than one repo in repo view...even for a minimal workspace. OTOH I think the notion of having just one repo might be nice for the 'minimalist' dev but it also might confuse some users to assume that there is only one repo _possible_.
> 
> As said, it is subjective but I assigned this to me and will make a proposal.

As I said to begin with, the org.osgi.annotation.bundle:2.0.0 bundle is needed for the new bndtools osgi service wizard to function.   Please at least put that in the minimal workspace template, as otherwise I believe it's going to cripple both the service wizard and the minimal workspace template.   

But once again, I would suggest giving less credence to your own preferences (i.e. minimalism) and more to the needs and expectations of the non-bndtools developer coming to bndtools (e.g. usecases and expectations of non-osgi/non-bndtools devs).   If bndtools can't get more of them, eventually it's not going to matter how simple it seems to the OSGi experts.

**Comment by pkriens** [comment]
I am getting a bit lost and I moved it to be too philosophical so let me try to recap.

The `github.com/bndtools/workspace` is the default and we try to make that complete for an out of the box experience. The purpose of the `github.com/bndtools/bndtools.workspace.min` was to have an absolute minimum workspace with the idea to drop files in the `cnf/ext` folder for extensions. The [`-library`](https://bnd.bndtools.org/instructions/library.html) is also very suitable for this.

If we keep this workspace minimal, we could make files for `cnf/ext` or libraries that people then could plugin? I.e. we could have a OSGi 8/Eclipse plugin or a OSGi 8/Felix. 

Is there a way to join the call next Friday? If not, can we setup a call to discuss this with a group of people? I think Data in Motion has very similar requirements and they seem to like the library idea.

**Comment by scottslewis** [comment]
> I am getting a bit lost and I moved it to be too philosophical so let me try to recap.
> 
> The `github.com/bndtools/workspace` is the default and we try to make that complete for an out of the box experience. The purpose of the `github.com/bndtools/bndtools.workspace.min` was to have an absolute minimum workspace with the idea to drop files in the `cnf/ext` folder for extensions. The [`-library`](https://bnd.bndtools.org/instructions/library.html) is also very suitable for this.
> 
> If we keep this workspace minimal, we could make files for `cnf/ext` or libraries that people then could plugin? I.e. we could have a OSGi 8/Eclipse plugin or a OSGi 8/Felix.

My point is simply this:   If you have something that's so minimal that it's useless for *everyone* ( (as it was) .   Then why have it?   I've tried to use it to start a small/minimal project several times, and spent so much time adding to it/extending it to allow *any* usage that I've ended up giving up and never using it (rather always starting from a more capable workspace and adding/removing stuff).   My guess/believe is that most people coming to bndtools likely have a similar experience and perhaps never use bndtools again to do even the most trivial example, they have to learn 20 new concepts about bndtools (eclipse), and then configure the tooling themselves.

My position is:  Either make it useful for *something/any real use case* or remove it.   My minimum proposal for use case is the new osgi service wizards I've contributed...which require org.osgi.annotation.bundle:2.0.0 to build/compile.    

Without anything else from repos (api/framework impl, ds, etc) users won't be able to start/run the new osgi service projects as they have a bndrun file (which obviously can't resolve if no framework is available).   It's frequently very helpful for tooling newbies to be able to go from nothing to functioning/running code and I've tried to accommodate that with the Java service wizard.

> 
> Is there a way to join the call next Friday? If not, can we setup a call to discuss this with a group of people? I think Data in Motion has very similar requirements and they seem to like the library idea.

I'm not sure which Friday you are referring to...i.e. tomorrow/1 or next week/8?   I don't know the time but have real work in Pacific time work hours.   I'll consider it if you let me know date and time and I can make it.

**Comment by pkriens** [comment]
The call is next week Friday, June 9.

**Comment by scottslewis** [comment]
@pkriens Unfortunately I can't make the bndtools call/meeting on coming Friday.   I have, however, reduced the number of repos used in this pull request to 3...i.e. OSGi API, OSGi Impl and Maven Central (with only the org.osgi.annotation.bundle in central.maven file).   I've taken out everything else except the -contract: * from build.bnd.

I couldn't figure out where the R8 API and R8 Ref Impl bundles are so I left things at R7.   Note also that bndtools.workspace is also currently at R7 and if you move this one (minimal) to 8 they probably should update everything in bndtools.workspace as well.
